---
title: "Genetic Diversity"
author: "Paulo Manrique"
date: "2023-05-27"
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
    code_folding: show
editor_options: 
  markdown: 
    wrap: 72
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Background

```{r}
setwd('~/Documents/Github/Plasmodium_WGS_analysis/GeneticDiversity/')
load('ColombiaGates_filtered_rGenome.RData')
source('../functions_libraries/load_libraries.R')
source('../functions_libraries/functions.R')
sourceCpp('../functions_libraries/Rcpp_functions.cpp')
```

## adding metadata

```{r}
external_metadata = read.csv('merged_metadata.csv')

# merge data bases

rGenome_object@metadata = merge(rGenome_object@metadata, external_metadata[,c("Sample_id", "Study",'Date_of_Collection', "Year_of_Collection",'Subnational_level0', 'Subnational_level1', "Subnational_level2", "ParasiteCount", "Parasitemia", "Ethnicity")], by = 'Sample_id', all.x = T)

# Sorting samples
rownames(rGenome_object@metadata) = rGenome_object@metadata$Sample_id
rGenome_object@metadata = rGenome_object@metadata[colnames(rGenome_object@gt),]

# impute missing Subnational_level0

rGenome_object@metadata[is.na(rGenome_object@metadata$Subnational_level0),]

rGenome_object@metadata %<>% mutate(Subnational_level0 = case_when(
  is.na(Subnational_level0) & (Subnational_level1 %in% c('Valle del Cauca',
                                                         'Narino',
                                                         'Choco')) ~ 'Pacific Coast',
  
  is.na(Subnational_level0) & Subnational_level1 == 'Cordoba' ~ 'North',
  
  !is.na(Subnational_level0) ~ Subnational_level0
))

```


## Handel Polyclonal infections

Create two objects containing the list of samples that are monoclonals and polyclonals
```{r}
monoclonals = rGenome_object@metadata %>% 
  filter(Clonality %in% c('Monoclonal')) %>%
  select(Sample_id) %>% unlist()

polyclonals = rGenome_object@metadata %>% 
  filter(Clonality %in% c('Polyclonal', 'Highly related Polyclonal')) %>%
  select(Sample_id) %>% unlist()
```

Create a new rGenome object that only has the information of monoclonals samples
```{r}
rGenome_object_mono = filter_samples(
      rGenome_object , 
      v = rGenome_object@metadata$Sample_id %in% monoclonals)
```
Calculate the expected heterozygosity assuming that all samples are diploids
```{r}
rGenome_object@loci_table$ExpHet_all_as_diploid = 
  get_ExpHet(obj = rGenome_object, update_AC = T, monoclonals = NULL, polyclonals = c(monoclonals, polyclonals))
```

Calculate the expected heterozygosity considering that all samples are monoclonal, it means that secondary alleles won't be included
```{r}
rGenome_object@loci_table$ExpHet_all_as_haploid = 
  get_ExpHet(obj = rGenome_object, update_AC = T, monoclonals = c(monoclonals, polyclonals))
```

Calculate the expected heterozygosity considering only monoclonal samples
```{r}
rGenome_object@loci_table$ExpHet_only_monoclonals = 
  get_ExpHet(
    rGenome_object_mono, 
    update_AC = T, 
    monoclonals = monoclonals)
```

Calculate the expected heterozygosity defining the ploidy of each position individually
```{r}
rGenome_object@loci_table$ExpHet_default = get_ExpHet(obj = rGenome_object)
```

Calculate the expected heterozygosity defining the ploidy of each samples based on our previous classification. It means that in monoclonal samples secondary alleles will be removed, while in polyclonal samples all observed alleles will be included
```{r}
rGenome_object@loci_table$ExpHet_default2 = 
  get_ExpHet(obj = rGenome_object, update_AC = T, monoclonals = monoclonals, polyclonals = polyclonals)
```

Create a plot of ExpHet using the different approaches
```{r}
rGenome_object@loci_table %>%
  pivot_longer(cols = c('ExpHet_all_as_haploid', 'ExpHet_only_monoclonals', 'ExpHet_default', 'ExpHet_default2'), names_to = 'Treatment', values_to = 'ExpHet')%>%
  ggplot(aes(x = ExpHet_all_as_diploid, y = ExpHet))+
  geom_point(alpha = .1)+
  facet_wrap(.~Treatment, ncol = 4)+
  theme_bw()

```
Create a plot of the distribution of ExpHet by by type of marker
```{r}
rGenome_object@loci_table %>%
  pivot_longer(cols = c('ExpHet_all_as_diploid','ExpHet_all_as_haploid', 'ExpHet_only_monoclonals', 'ExpHet_default', 'ExpHet_default2'), names_to = 'Treatment', values_to = 'ExpHet')%>%
  ggplot(aes(fill = Treatment, x = ExpHet))+
  geom_histogram(alpha = .4)+
  facet_grid(TypeOf_Markers~Treatment, scales = 'free_y')+
  theme_bw()

```
Calculate Heterozygosity by each population (Subnational_level0: Regions within Colombia)
```{r}
ExpHet_by_Pop = 
  get_ExpHet(obj = rGenome_object, update_AC = T, monoclonals = monoclonals, polyclonals = polyclonals, by = 'Subnational_level0')
```

```{r}
ExpHet_by_Pop %>%
  pivot_longer(cols = c('North', 'Pacific Coast'), names_to = 'Region', values_to = 'ExpHet')%>%
  ggplot(aes(fill = Region, x = ExpHet))+
  geom_histogram(alpha = .4)+
  facet_grid(TypeOf_Markers~Region, scales = 'free_y')+
  theme_bw()

```

## Nucleotide diversity

Calculate Nuc. Div. by gene
```{r}
pi_by_gene = get_nuc_div(rGenome_object, monoclonals = monoclonals, polyclonals = polyclonals, gff = '../reference/genes.gff', type_of_region = 'gene', window = NULL, by = NULL)
```

Calculate Nuc. Div. by gene by Population (Subnational_level0)
```{r}
pi_by_gene_by_snl0 = get_nuc_div(rGenome_object, monoclonals = monoclonals, polyclonals = polyclonals, gff = '../reference/genes.gff', type_of_region = 'gene', window = NULL, by = 'Subnational_level0')
```

Calculate Nuc. Div. by window
```{r}
pi_ws_10000 = get_nuc_div(rGenome_object, monoclonals = monoclonals, polyclonals = polyclonals, gff = '../reference/genes.gff', type_of_region = 'gene', window = 10000, by = NULL)
```

Calculate Nuc. Div. by window by Population (Subnational_level0)
```{r}
pi_ws_10000_by_snl0 = get_nuc_div(rGenome_object, monoclonals = monoclonals, polyclonals = polyclonals, gff = '../reference/genes.gff', type_of_region = 'gene', window = 10000, by = 'Subnational_level0')
```

```{r}

mhp_pi_by_gene_by_snl0 = pi_by_gene_by_snl0 %>%
  pivot_longer(cols = c('North_pi', 'Pacific Coast_pi', 'Total_pi'), names_to = 'Regions', values_to = 'pi')%>%
  mutate(CHROM2  = gsub('(^(\\d|P|v)+_|_v1)', '',seqid))%>%
  ggplot(aes(x = start, y = pi, color = Regions)) +
    geom_point(alpha = 0.5, size = .25) +
    scale_color_manual(values = c('dodgerblue3', 'firebrick3', 'gray40'))+
    facet_grid(Regions ~ CHROM2, scales = 'free_x')+
    theme_minimal()+
    labs(y = 'Nuc. Div.', x = 'Chromosomal position', color = 'Region')+
    theme(legend.position = 'right',
          axis.text.x = element_blank(),
          axis.title.x = element_blank())

mhp_pi_by_gene_by_snl0
```

```{r}

mhp_pi_by_gene_by_snl0 = ggplotly(mhp_pi_by_gene_by_snl0)

chromosomes = c('PvP01_01_v1', 'PvP01_02_v1', 'PvP01_03_v1', 'PvP01_04_v1',
                'PvP01_05_v1', 'PvP01_06_v1', 'PvP01_07_v1', 'PvP01_08_v1',
                'PvP01_09_v1', 'PvP01_10_v1', 'PvP01_11_v1', 'PvP01_12_v1',
                'PvP01_13_v1', 'PvP01_14_v1', 'PvP01_MIT_v1')

for(region in 1:3){
  for(chrom in 1:16){
    mhp_pi_by_gene_by_snl0$x$data[[chrom + (region - 1)*15]]$text = paste(mhp_pi_by_gene_by_snl0$x$data[[chrom + (region - 1)*15]]$text,
      pi_by_gene_by_snl0[pi_by_gene_by_snl0$seqid == chromosomes[chrom],]$gene_description, sep = '<br />')
  }
}


mhp_pi_by_gene_by_snl0

```

## Genetic differentiation

```{r}

install.packages('RcppEigen')
library(RcppEigen)

sourceCpp('../functions_libraries/Rcpp_functions.cpp')

pca_colombia = fastGRM(obj = rGenome_object, k = 2, monoclonals = monoclonals, polyclonals = polyclonals, Pop = 'Subnational_level0')


Plot_PCs_COL_WGS = pca_colombia  %>% ggplot(aes(x = PC1, y = PC2, color = Subnational_level0))+
  geom_point(alpha = .7, size = 2) +
  stat_ellipse(level = .6)+
  scale_color_manual(values = c('dodgerblue3', 'firebrick3'))+
  theme_bw()+
  labs(title = 'WGS - Colombia',
       color = 'Regions')

```